### ISSUE TITLE
How to test a function that using knex?

### ISSUE THREAD

**Issue by jon9090** [knex/knex#5475]
I have this function and I want to write a test.
```
const getLastUpdatedUsers = (date) => {
 const users = knex('users').select('id', 'name').where({ update_at < date });
 return users;
}
```

How do I write a test for this function? do I mock the knex? do I have to connect to the "test database" and run the actual SQL query? 

Basically, I want to make sure I write my knex query right. this is mean if the next developer change to `update_at > date ` it should cause an error in tests.

In knex, I build some logic (`where`) and I want to test it. How to do it?

**Comment by Naddiseo** [comment]
If you want to test the actual sql query you'll need a test database. [knex's own test suite](https://github.com/knex/knex/tree/master/test) has integration testing similar to what you're looking for.

**Comment by Naddiseo** [comment]
>  to do that I need to create a mock database right

Yes, don't use a production database with actual data for testing. The [knexfile](https://knexjs.org/guide/migrations.html#environment-configuration) allows for different configs depending on which environment you're in, you can create a separate `test` environment.

> each entity is connected to different tables and the data is about 1T. so I can't use create table or drop table.

Usually with integration testing you create test data specifically for what you're testing  during the test itself. If you need to clear data state between tests, I've found truncating the tables to be fastest. Integration testing tends to be very slow though.

> is it possible to run knex command to get all the schema and data from the database and use it in memory?

Your test database should be just the schemas, custom types, triggers etc, but no data from production. You should be able to use your existing knex migration files to create the test database.

**Comment by mercmobily** [comment]
Hi! Thanks for asking — there are a few reliable patterns.

Treat Knex as a dependency to inject/stub. For unit tests, pass a `knex` instance configured for a test DB (for example SQLite memory) or supply an object with the subset of methods you call (for example `select/insert/update/raw`) and stub their behaviour. For integration tests, use a real `knex` connected to a test database and migrate/seed in a `beforeAll`, tear down in `afterAll`.

In the codebase we use standard patterns; there is no magical singleton. You can create and dispose of instances freely, and hooks like `postProcessResponse` and `wrapIdentifier` allow customisation if needed.

Prefer dependency injection: export your function as `(db) => async (...) => { /* use db(...) */ }` so tests can supply a mock or test instance. If you share a small snippet, we can sketch the exact test setup.

Closing for now since the issue is quite old.
